### PR TITLE
Introduce trait StakeTableScheme and migrate previous implementations

### LINKED ISSUE #65: Add trait for Stake Table
As a followup of https://github.com/EspressoSystems/hotshot-primitives/pull/64#issuecomment-1601905381, we will define the `trait StakeTable` specified in the hotshot spec (with 2 minor pointers for slight modifications in https://github.com/EspressoSystems/papers/issues/215).

### PR DESCRIPTION
**Pull request by alxiong**
Resolve #65. 
(a follow-up of, and should be merged after, #64)

Major changes include:
- introduce `trait StakeTableScheme`
- rename `SnapshotVersion` from `pending/frozen/active` to `head/epoch_start/last_epoch_start` for better readability
- most old APIs are migrated under the trait impl, notable additions include: `len()`, `contains_key()`, `deregister()` (left unimplemented and future work); notable changes include: `lookup()` now returns `(Amount, Proof)` instead of just `MerkleProof` as before.
- [x] replace `struct StakeTableEntry<V>` with generic API from `trait StakeTableScheme`
  - [x] implement an Iterator for `PersistentMerkleNode`, implement `IntoIterator` for `StakeTableScheme` 
  - [x] Make current stake_table implementation generic over `K` instead of fixated on `EncodedPublicKey`

### CODE REVIEW

**Comment by alxiong** [comment]
Q: should we make `struct StakeTable` generic over `<K, V>`?

currently our stake table implementation uses concrete types `(EncodedPublicKey, U256)` for its key-value, where the encoded pk just contains the seralized bytes (of whatever actual pk being used)

This decision forces me to write: 
```rust
/// Public parameters for [`BitVectorQC`].
/// generic over the stake table type `ST` and aggregated signature type `A`
#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub struct QCParams<A: AggregateableSignatureSchemes, ST: StakeTableScheme> {
    // pub stake_entries: Vec<StakeTableEntry<V>>,
    pub stake_table: Arc<ST>,
    pub threshold: U256,
    pub agg_sig_pp: A::PublicParameter,
}

impl<A, ST> QuorumCertificate<A> for BitVectorQC<A, ST>
where
    A: AggregateableSignatureSchemes + Serialize + for<'a> Deserialize<'a>,
    ST: StakeTableScheme<Amount = U256>,
    ST::Key: From<A::VerificationKey>,
    A::VerificationKey: From<ST::Key>,
```

instead of

```rust
impl<A, ST> QuorumCertificate<A> for BitVectorQC<A, ST>
where
    A: AggregateableSignatureSchemes + Serialize + for<'a> Deserialize<'a>,
    ST: StakeTableScheme<Key = A::VerificationKey>,
    ST::Amount: Add<Output = Self>,
```

Apart from having cleaner trait bound, the main advantages of the latter (if we have generic `struct StakeTable<K, V>`) is avoiding lots of internal serde. 
For example, inside`QC::check()`, we will be calling `A::multi_sig_verify` on the aggregated signature scheme which accepts list of public keys of type `&[A::VerificationKey]`, which may not be of the same type as `&[ST::Key]` (which currently is concretized as `struct EncodedPublicKey`).
To deal with this, we would have to implement two-way conversion (thus the bound `A::VerificationKey: From<ST::Key>`), and this (arguably unnecessary) serde is being done each time people run `QC::check()`. 


In short, I would propose that we make turn our Stake Table generic `struct StakeTable<K, V>`, wdyt? @mrain 
(if agreed, I would work on a separate PR that merge into this PR)

**Comment by alxiong** [comment]
Here's a problem: 

previously, we had this VerKey -> ark_bn254::Fq conversion:
https://github.com/EspressoSystems/hotshot-primitives/blob/e9bda5ca1582c957c27e98b7ccf96b7a80dff13e/src/stake_table/mod.rs#L240

but this code has the wrong assumption that `bls_on_bn254::VerKey` can be serialized into 32 bytes without any loss/truncation. 

`VerKey` is of type `G2Affine` which has x, y coordinate where each is a quadratic extension field, thus at least even with compressed serialized bytes, we need 64 bytes. 
And lossy conversion would result in failed deserialization back into the correct VerKey. 

---

at the current stage of this PR, I introduce this `IntoField<FieldType>` trait bound, as a reflection of the old code (thus the wrong assumption that it can be converted to a `FieldType` w/o loss),

And I think I might be better refactoring it to `IntoFields<FieldType; const N>` and then change:
https://github.com/EspressoSystems/hotshot-primitives/blob/e9bda5ca1582c957c27e98b7ccf96b7a80dff13e/src/stake_table/utils.rs#L94-L100
and 
https://github.com/EspressoSystems/hotshot-primitives/blob/e9bda5ca1582c957c27e98b7ccf96b7a80dff13e/src/stake_table/config.rs#L13-L14

into maybe `type Digest<N> = FixedLengthRescueCRHF<FieldType, TREE_BRANCH + N-1, 1>`, 
I'm not sure if I could make the compiler happy, but this seems a necessary change.

---
There's another almost hacky way to avoid deserialization error. that is to define and manually implement serialization of `struct StakeTable` as the its commitment, meaning I don't serialize the whole tree (thus avoiding serializing the `Key` type), instead I only serialize the commitment.

^^ the problem with this approach is given a serialized `QCparam`, you can't just directly reconstruct the stake table. instead you can only check for consistency against another stake table you download from other means. <-- which also seem a little strange. 


@mrain let's chat and see which one do we prefer and what to do next.

**Comment by alxiong** [comment]
After discussion with @mrain, we decide to go with: `IntoFields<[FieldType; 2]>`.
While we acknowledge that this looks hardcoded, we prefer it over using `VariableLengthRescueCRHF` which will automatically append an 1, and ends up costing 2 rescue hash instead of 1 for the `Digest::evaluate()`.

File: src/stake_table/mod.rs [hunk 1]
```diff
@@ -1,175 +1,312 @@
...
+
+    /// Register a new key into the stake table.
+    fn register(&mut self, new_key: Self::Key, amount: Self::Amount)
+        -> Result<(), StakeTableError>;
+
+    /// Batch register a list of new keys. A default implementation is provided
+    /// w/o batch optimization.
+    fn batch_register(
+        &mut self,
+        new_keys: Vec<Self::Key>,
```
Comment (mrain) [inline src/stake_table/mod.rs:51]: Why taking `Vec` as input, not using slice or maybe `IntoIter`?

**Review by mrain — CHANGES_REQUESTED** [review]
Generally LGTM, with some minor comments

File: src/stake_table/mod.rs (hunk 1 above)
Comment (alxiong) [reply src/stake_table/mod.rs:51]: Cuz register() takes ownership of Key, so I mimic the pattern and taking `Vec<Key>`. 
So maybe change to `IntoIter`? what about other APIs like `batch_update` that takes slices?

Comment (alxiong) [reply src/stake_table/mod.rs]: done in [e92fb9b](https://github.com/EspressoSystems/hotshot-primitives/pull/66/commits/e92fb9b76a6e407af8371dc7e061ff6778f5d796)

File: src/stake_table/mod.rs [hunk 2]
```diff
@@ -1,175 +1,312 @@
...
+
+    /// Returns the number of keys in the `version` of the table.
+    fn len(&self, version: SnapshotVersion) -> Result<usize, StakeTableError>;
+
+    /// Returns true if `key` is currently registered, else returns false.
+    fn contains_key(&self, key: &Self::Key) -> bool;
+
+    /// Lookup the stake under a key against a specific historical `version`,
+    /// returns error if keys unregistered.
+    fn lookup(
```
[the remark this answers is not in the document]
Comment (mrain) [reply src/stake_table/mod.rs:93]: I vaguely remember that this is request by @jbearer , where light client may be sensitive to this extra cost.

[the remark this answers is not in the document]
Comment (alxiong) [reply src/stake_table/mod.rs:93]: done in [d8474ce](https://github.com/EspressoSystems/hotshot-primitives/pull/66/commits/d8474ce4156bfcb9c8e742faeb0e66cd723993f4)

File: src/stake_table/mod.rs [hunk 3]
```diff
@@ -1,198 +1,268 @@
...
+    type Amount: Clone + Copy;
+    /// type for the commitment to the current stake table
+    type Commitment;
+    /// type for the proof associated with the lookup result (if any)
+    type LookupProof;
+    /// type for the iterator over (key, value) entries
+    type IntoIter: Iterator<Item = (Self::Key, Self::Amount)>;
+
+    /// Register a new key into the stake table.
+    fn register(&mut self, new_key: Self::Key, amount: Self::Amount)
```
Comment (mrain) [inline src/stake_table/mod.rs:44]: Oh you remind me, since `Key` only requires `Clone`, should we use `&Self::Key` here and make a clone inside? Or we could also require `Copy`.

**Review by mrain — APPROVED** [review]
